File: longestPalindromeByConcatenating2LetterWords.go
func longestPalindrome(words []string) int {
    count := make(map[string]int)
    for _, word := range words {
        count[word]++
    }

    length := 0
    centralUsed := false

    for word, c := range count {
        rev := string([]byte{word[1], word[0]})

        if word == rev {
            // Pair palindromic words like "gg"
            pairs := c / 2
            length += pairs * 4
            count[word] -= pairs * 2
            if !centralUsed && count[word] > 0 {
                length += 2
                centralUsed = true
            }
        } else if word < rev { // avoid double counting
            if revCount, ok := count[rev]; ok {
                pairs := min(c, revCount)
                length += pairs * 4
                count[word] -= pairs
                count[rev] -= pairs
            }
        }
    }

    return length
}

func min(a, b int) int {
    if a < b {
        return a
    }
    return b
}

File: longestPalindromeByConcatenating2LetterWords.py
from collections import Counter
from typing import List

class Solution:
    def longestPalindrome(self, words: List[str]) -> int:
        count = Counter(words)
        length = 0
        central_used = False

        for word in list(count.keys()):
            reversed_word = word[::-1]

            if word == reversed_word:
                # Pair up same-letter words like "gg"
                pairs = count[word] // 2
                length += pairs * 4
                count[word] -= pairs * 2
                if not central_used and count[word] > 0:
                    length += 2  # one central word
                    central_used = True
            else:
                if reversed_word in count:
                    pairs = min(count[word], count[reversed_word])
                    length += pairs * 4
                    count[word] -= pairs
                    count[reversed_word] -= pairs

        return length
